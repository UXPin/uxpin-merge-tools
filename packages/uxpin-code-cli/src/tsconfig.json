{
  "compilerOptions": {
    "target": "es6",
    "module": "commonjs",
    "sourceMap": true,
    "strict": true,
    "moduleResolution": "node",
    "baseUrl": "../",
    "skipLibCheck": true,
    "outDir": "../dist/",
    "paths": {
      "*": [
        "node_modules/*",
        "src/types/*"
      ]
    }
  }
}