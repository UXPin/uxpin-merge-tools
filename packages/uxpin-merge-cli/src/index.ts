import 'source-map-support/register';
export { runProgram } from './program/runProgram';
export { Command } from './program/command/Command';
export { getToolVersion } from './program/utils/version/getToolVersion';
export { buildDesignSystemWithStorybook } from './steps/building/buildDesignSystemWithStorybook';
