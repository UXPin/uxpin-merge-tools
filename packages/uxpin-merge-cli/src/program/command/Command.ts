export enum Command {
  DUMP = 'dump',
  EXPERIMENT = 'experiment',
  INIT = 'init',
  GENERATE_PRESETS = 'generate-presets',
  PUSH = 'push',
  DELETE_VERSION = 'delete-version',
  SERVER = 'server',
  SUMMARY = 'summary',
}

export const DEFAULT_COMMAND: Command = Command.EXPERIMENT;
