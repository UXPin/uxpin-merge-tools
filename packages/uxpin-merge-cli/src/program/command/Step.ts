import { DSMetadata } from '../DSMeta';

export interface Step {
  shouldRun: boolean;
  exec: (infos: DSMetadata) => any;
}

export type StepExecutor = (designSystem: DSMetadata) => Promise<DSMetadata>;

export type StepWithoutDSExecutor = () => Promise<void>;
