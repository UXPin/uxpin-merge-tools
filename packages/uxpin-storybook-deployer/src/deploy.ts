import { PresignedPost } from 'aws-sdk/clients/s3';
import axios, { AxiosError, AxiosPromise, AxiosRequestConfig, AxiosResponse } from "axios";
import { execSync } from 'child_process';
import { createReadStream, readdirSync, statSync } from 'fs';
import { relative, join } from 'path';
import FormData from 'form-data';
import { IncomingMessage } from 'http';
import chalk from 'chalk';
import mime from 'mime-types';

type DeployOptions = {
  source: string,
  token: string,
  uxpinDomain?: string,
}

type GetLibraryHashByTokenResponse = {
  libraryHash: string,
}

type UpdateCommitHashBody = {
  commitHash: string,
}

const UXPIN_DOMAIN = 'uxpin.com';
const AUTH_TOKEN_HEADER_NAME = 'auth-token';

let latestCommitHash:string;

export async function deploy(args: DeployOptions):Promise<void> {
  try {
    latestCommitHash = getLatestCommitHash(args.source);
    const { libraryHash } = await getLibraryHashByToken(args);
    const presignedS3PostUrlResponse = await getPresignedS3PostUrl(args);
    await deployStorybookToS3(args, libraryHash, presignedS3PostUrlResponse);
    console.log('Updating commit hash on UXPin...');
    await updateCommitHash(args);
    const storybookUrl = join(getBaseUrl(args.uxpinDomain), 'libraries-storybook', libraryHash, 'index.html');
    console.log(chalk.green(`âœ… DONE! Your storybook is available on ${storybookUrl}`));
  } catch (error) {
    typeof(error) === 'string' ? console.log(chalk.red(error)) : console.log(error);
  }
}

function getBaseUrl(uxpinDomain:string | undefined):string {
  return `https://api.${uxpinDomain || UXPIN_DOMAIN}`;
}

async function getLibraryHashByToken(args:DeployOptions):Promise<GetLibraryHashByTokenResponse> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/getLibraryHashByToken`;
  return axiosWithEnhancedError({ url, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token, method: 'GET' }})
    .then((res:AxiosResponse) => res.data as GetLibraryHashByTokenResponse);
}

async function getPresignedS3PostUrl(args:DeployOptions):Promise<PresignedPost> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/getPresignedS3PostUrl/${latestCommitHash}`;
  return axiosWithEnhancedError({ url, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token, method: 'GET' }})
    .then((res:AxiosResponse) => res.data as PresignedPost);
}

function getLatestCommitHash(source:string):string {
  try {
    return execSync(`cd ${source} && git rev-parse HEAD`).toString().trim();
  } catch (error) {
    throw `ERROR: Failed to get commit hash from ${source}. Please make sure you can run\ngit rev-parse HEAD \ncommand in the directory because UXPin requires latest commit hash`;
  }
}

function getS3Prefix(libraryHash:string):string {
  return `${libraryHash}/${latestCommitHash}/`;
}

function getFilesList(source:string):string[] {
  let results:string[] = [];
  const list:string[] = readdirSync(source);
  list.forEach((fileOrDir) => {
      fileOrDir = source + '/' + fileOrDir;
      const stat = statSync(fileOrDir);
      if (stat && stat.isDirectory()) {
        results = results.concat(getFilesList(fileOrDir));
      } else {
        results.push(fileOrDir);
      }
    });
  return results;
}

async function deployStorybookToS3(args:DeployOptions, libraryHash:string, presignedS3urlResponse:PresignedPost):Promise<void[]> {
  console.log('Start uploading Storybook assets...');
  const s3Prefix = getS3Prefix(libraryHash);
  const promises:Array<Promise<void>> = getFilesList(args.source).map(async (absFilePath) => {
    const relativePath = relative(args.source, absFilePath);
    const key = join(s3Prefix, relativePath);

    const form = new FormData();
    Object.entries(presignedS3urlResponse.fields).forEach(([field, value]) => {
      form.append(field, value);
    });
    form.append('key', key);
    form.append('Content-Type', mime.contentType(mime.lookup(absFilePath) || 'application/octet-stream'));
    form.append('file', createReadStream(absFilePath));

    return new Promise((resolve, reject) => form.submit(presignedS3urlResponse.url, (err, data:IncomingMessage) => {
      if (err) {
        reject(err);
      }
      else if (data.statusCode !== 204) {
        reject({
          status: data.statusCode,
          statusMessage: data.statusMessage,
          filePath: relativePath,
        });
      }
      resolve();
    }));
  });

  return Promise.all(promises);
}

async function updateCommitHash(args:DeployOptions):Promise<boolean> {
  const url = `${getBaseUrl(args.uxpinDomain)}/libraries-storybook/updateCommitHash`;
  const data:UpdateCommitHashBody = { commitHash: latestCommitHash };
  return axiosWithEnhancedError({ url, data, headers: { [AUTH_TOKEN_HEADER_NAME]: args.token }, method: 'PUT'})
    .then(() => true);
}

function axiosWithEnhancedError(options:AxiosRequestConfig):AxiosPromise {
  return axios(options).catch((error:AxiosError) => {
    if (error.response) {
      throw {
        ...error.response.data,
        url: options.url,
      };
    }
    throw error;
  });
}
