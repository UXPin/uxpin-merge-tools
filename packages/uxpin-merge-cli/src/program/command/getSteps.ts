import { ProgramArgs } from '../args/ProgramArgs';
import { Command } from './Command';
import { getDeleteVersionCommandSteps } from './delete_version/getDeleteVersionCommandSteps';
import { getDumpCommandSteps } from './dump/getDumpCommandSteps';
import { getExperimentationCommandSteps } from './experimentation/getExperimentationCommandSteps';
import { getGeneratePresetsCommandSteps } from './generate_presets/getGeneratePresetsCommandSteps';
import { getInitCommandSteps } from './init/getInitCommandSteps';
import { getPushCommandSteps } from './push/getPushCommandSteps';
import { getServerCommandSteps } from './server/getServerCommandSteps';
import { Step } from './Step';
import { getSummaryCommandSteps } from './summary/getSummaryCommandSteps';

export function getSteps(args: ProgramArgs): Step[] {
  switch (args.command) {
    case Command.DUMP:
      return getDumpCommandSteps();

    case Command.EXPERIMENT:
      return getExperimentationCommandSteps(args);

    case Command.INIT:
      return getInitCommandSteps(args);

    case Command.GENERATE_PRESETS:
      return getGeneratePresetsCommandSteps(args);

    case Command.PUSH:
      return getPushCommandSteps(args);

    case Command.DELETE_VERSION:
      return getDeleteVersionCommandSteps(args);

    case Command.SERVER:
      return getServerCommandSteps(args);

    case Command.SUMMARY:
      return getSummaryCommandSteps();

    default:
      return [];
  }
}
