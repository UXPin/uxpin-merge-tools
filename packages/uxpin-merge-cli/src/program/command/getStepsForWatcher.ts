import { ProgramArgs } from '../args/ProgramArgs';
import { Command } from './Command';
import { getExperimentationWatchCommandSteps } from './experimentation/getExperimentationWatchCommandSteps';
import { Step } from './Step';

export function getStepsForWatcher(args: ProgramArgs): Step[] {
  if (args.command === Command.EXPERIMENT) {
    return getExperimentationWatchCommandSteps(args);
  }

  return [];
}
