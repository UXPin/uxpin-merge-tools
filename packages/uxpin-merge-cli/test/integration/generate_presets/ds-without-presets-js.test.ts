import { readFile, rmdirSync, unlinkSync } from 'fs-extra';
import { Command } from '../../../src';
import { runUXPinMergeCommand } from '../../utils/command/runUXPinMergeCommand';
import { setTimeoutBeforeAll } from '../../utils/command/setTimeoutBeforeAll';

const CURRENT_TIMEOUT = 30000;
setTimeoutBeforeAll(CURRENT_TIMEOUT);

const DS_DIR = 'test/resources/designSystems/withoutPresetsJs';
describe('The generate presets command', () => {
  describe('run for Avatar component', () => {
    const COMPONENT_DIR = 'src/components/Avatar';

    afterEach(() => {
      unlinkSync(`${DS_DIR}/${COMPONENT_DIR}/presets/0-default.jsx`);
      rmdirSync(`${DS_DIR}/${COMPONENT_DIR}/presets`);
    });

    it('create 0-default.jsx', () => {
      // when
      return runUXPinMergeCommand({
        cwd: 'resources/designSystems/withoutPresetsJs',
        params: [Command.GENERATE_PRESETS, `--component-path "${COMPONENT_DIR}/Avatar.jsx"`],
      }).then(async () => {
        // then
        const content: string = await readFile(`${DS_DIR}/${COMPONENT_DIR}/presets/0-default.jsx`, 'utf-8');
        expect(content).toMatchSnapshot();
      });
    });
  });

  describe('run for Button component', () => {
    const COMPONENT_DIR = 'src/components/Button';

    afterEach(() => {
      unlinkSync(`${DS_DIR}/${COMPONENT_DIR}/presets/0-default.jsx`);
      rmdirSync(`${DS_DIR}/${COMPONENT_DIR}/presets`);
    });

    it('create 0-default.jsx', () => {
      // when
      return runUXPinMergeCommand({
        cwd: 'resources/designSystems/withoutPresetsJs',
        params: [Command.GENERATE_PRESETS, `--component-path "${COMPONENT_DIR}/Button.jsx"`],
      }).then(async () => {
        // then
        const content: string = await readFile(`${DS_DIR}/${COMPONENT_DIR}/presets/0-default.jsx`, 'utf-8');
        expect(content).toMatchSnapshot();
      });
    });
  });

  describe('run for whole ds based on uxpin.config.js', () => {
    const COMPONENT_DIRS: string[] = ['src/components/Avatar', 'src/components/Button'];

    afterAll(async () => {
      COMPONENT_DIRS.forEach((dir) => {
        unlinkSync(`${DS_DIR}/${dir}/presets/0-default.jsx`);
        rmdirSync(`${DS_DIR}/${dir}/presets`);
      });
    });

    beforeAll(async () => {
      await runUXPinMergeCommand({
        cwd: 'resources/designSystems/withoutPresetsJs',
        params: [Command.GENERATE_PRESETS, `--config "uxpin.config.js"`],
      });
    });

    it('create 0-default.jsx for Avatar', async () => {
      // when
      const content: string = await readFile(`${DS_DIR}/${COMPONENT_DIRS[0]}/presets/0-default.jsx`, 'utf-8');
      expect(content).toMatchSnapshot();
    });

    it('create 0-default.jsx for Button', async () => {
      // when
      const content: string = await readFile(`${DS_DIR}/${COMPONENT_DIRS[1]}/presets/0-default.jsx`, 'utf-8');
      expect(content).toMatchSnapshot();
    });
  });
});
