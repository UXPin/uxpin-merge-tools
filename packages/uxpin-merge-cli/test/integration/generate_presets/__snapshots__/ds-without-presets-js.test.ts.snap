// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The generate presets command run for Avatar component create 0-default.jsx 1`] = `
"import React from 'react';
import Avatar from '../Avatar';

export default (
  <Avatar
    uxpId=\\"avatar-1\\"
    size=\\"xs\\"
    imageUrl=\\"\\"
  />
);"
`;

exports[`The generate presets command run for Button component create 0-default.jsx 1`] = `
"import React from 'react';
import Button from '../Button';

export default (
  <Button
    uxpId=\\"button-1\\"
    primary={false}
  >Put Button contents here</Button>
);"
`;

exports[`The generate presets command run for whole ds based on uxpin.config.js create 0-default.jsx for Avatar 1`] = `
"import React from 'react';
import Avatar from '../Avatar';

export default (
  <Avatar
    uxpId=\\"avatar-1\\"
    size=\\"xs\\"
    imageUrl=\\"\\"
  />
);"
`;

exports[`The generate presets command run for whole ds based on uxpin.config.js create 0-default.jsx for Button 1`] = `
"import React from 'react';
import Button from '../Button';

export default (
  <Button
    uxpId=\\"button-1\\"
    primary={false}
  >Put Button contents here</Button>
);"
`;
